AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+        auto const m = MUST(input_stream.read_value<LittleEndian<u64>>());
+        v3 ^= m;
+
+        for (size_t i = 0; i < message_block_rounds; ++i)
+            SIPROUND;
+
+        v0 ^= m;
+    }
+
+    switch (left) {
+    case 7:
+        b |= (static_cast<u64>(input[block_end + 6])) << 48;
+        [[fallthrough]];
+    case 6:
+        b |= (static_cast<u64>(input[block_end + 5])) << 40;
+        [[fallthrough]];
+    case 5:
+        b |= (static_cast<u64>(input[block_end + 4])) << 32;
+        [[fallthrough]];
+    case 4:
+        b |= (static_cast<u64>(input[block_end + 3])) << 24;
+        [[fallthrough]];
+    case 3:
+        b |= (static_cast<u64>(input[block_end + 2])) << 16;
+        [[fallthrough]];
+    case 2:
+        b |= (static_cast<u64>(input[block_end + 1])) << 8;
+        [[fallthrough]];
+    case 1:
+        b |= (static_cast<u64>(input[block_end + 0]));
+        break;
+    case 0:
+        break;
+    }
+
+    v3 ^= b;
+
+    for (size_t i = 0; i < message_block_rounds; ++i)
+        SIPROUND;
+
+    v0 ^= b;
+
+    if (output.size() == 16)
+        v2 ^= 0xee;
+    else
+        v2 ^= 0xff;
+
+    for (size_t i = 0; i < finalization_rounds; ++i)
+        SIPROUND;
+
+    b = v0 ^ v1 ^ v2 ^ v3;
+
+    FixedMemoryStream output_stream { output };
+    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+
+    if (output.size() == 8)
+        return;
+
+    v1 ^= 0xdd;
+
+    for (size_t i = 0; i < finalization_rounds; ++i)
+        SIPROUND;
+
+    b = v0 ^ v1 ^ v2 ^ v3;
+    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+}
+
+// Using a shared key greatly simplifies DynamicLoader and Kernel,
+// since they don't need to worry about synchronizing access to block-scope statics.
+#ifdef KERNEL
+static u128 sip_hash_key = Kernel::get_good_random<u128>();
+#else
+// get_random is assumed to be secure, otherwise SipHash doesn't deliver on its promises!
+static u128 sip_hash_key = get_random<u128>();
+#endif
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+unsigned SipHash<message_block_rounds, finalization_rounds>::operator()(u64 input)
+{
+    ReadonlyBytes input_bytes { &input, sizeof(input) };
+    auto key = sip_hash_key;
+    u64 output;
+    Bytes output_bytes { &output, sizeof(output) };
+    do_siphash<message_block_rounds, finalization_rounds>(input_bytes, key, output_bytes);
+    return output;
+}
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+u64 sip_hash_bytes(ReadonlyBytes input)
+{
+    u64 output;
+    Bytes output_bytes { &output, sizeof(output) };
+    do_siphash<message_block_rounds, finalization_rounds>(input, sip_hash_key, output_bytes);
+    return output;
+}
+
+// Instantiate all used SipHash variants here:
+template struct SipHash<1, 3>;
+template struct SipHash<2, 4>;
+template struct SipHash<4, 8>;
+template u64 sip_hash_bytes<1, 3>(ReadonlyBytes);
+template u64 sip_hash_bytes<2, 4>(ReadonlyBytes);
```

Comment:
This overload is instantiated and never used outside a test (and any users would have to explicitly name it anyway), do consider just removing it.